R/readChromData.R

```diff
@@ -0,0 +1,246 @@
+#' but returns them. In other words, if three rows in one of the input
+#' `data.frame` contain the same values, the result `data.frame` also has the
+#' same 3 rows, while `unique` would collapse them into a single row. If rows
+#' with the same values are present multiple times in multiple input
+#' `data.frame`'s, e.g. 3 times in one `data.frame` and 2 times in
+#' another one, the result `data.frame` will contain the rows the maximum time
+#' they are present across tables. See examples for details.
+#'
+#' @note
+#'
+#' The ordering of a combined `data.frame` from `data.frame`s with some or
+#' all columns being `factors` might not be as expected. It is thus suggested
+#' to avoid passing `data.frame`s with `factors` to the function.
+#' 
+#' @param x
+#'
+#' @param cols `character` defining the columns that should be returned.
+#'
+#' @return `data.frame`
+#' 
+#' @noRd
+#'
+#' @md
+#' 
+#' @author Johannes Rainer
+#'
+#' @examples
+#'
+#' A <- data.frame(a = c(1, 1, 2, 3, 4, 5, 6), b = c(2, 2, 3, 4, 5, 5, 6))
+#' B <- data.frame(a = c(1, 3, 3, 3, 3, 4, 5), b = c(2, 4, 4, 4, 5, 5, 5))
+#' C <- data.frame(a = c(3, 3, 4, 4), b = c(4, 5, 5, 5))
+#'
+#' x <- list(A, B, C)
+#'
+#' .combine_data.frame(x)
+#'
+#' D <- data.frame(a = c("z", "b", "a", "g", "g"),
+#'                 b = c(1, 2, 3, 4, 4), stringsAsFactors = FALSE)
+#' E <- data.frame(a = c("g", "a", "d", "g"),
+#'                 b = c(4, 3, 1, 1), stringsAsFactors = FALSE)
+#' .combine_data.frame(list(D, E))
+.combine_data.frame <- function(x, cols) {
+    if (!length(x))
+        stop("length of 'x' must be > 0")
+    if (!all(unlist(lapply(x, is.data.frame))))
+        stop("all elements in 'x' need to be a data.frame")
+    fd <- do.call(rbind, x)
+    if (missing(cols))
+        cols <- colnames(fd)
+    else {
+        if (!all(cols %in% colnames(fd)))
+            stop("All columns specified with 'cols' have to be present in the",
+                 " data.frames")
+    }
+    ## Make first unique elements
+    fd <- unique(fd[, cols, drop = FALSE])
+    if (nrow(fd)) {
+        fd_id <- do.call(paste, fd)
+        ## Find then the number of times each should be repeated.
+        rep_each <- matrix(ncol = length(x), nrow = nrow(fd))
+        for (i in seq_along(x)) {
+            x_id <- do.call(paste, x[[i]][, cols, drop = FALSE])
+            cnts <- table(match(x_id, fd_id))
+            idx <- as.numeric(names(cnts))
+            rep_each[as.numeric(names(cnts)), i] <- as.numeric(cnts)
+        }
+        splt <- split.data.frame(fd, 1:nrow(fd))
+        fd <- do.call(rbind,
+                      rep(splt, apply(rep_each, MARGIN = 1, max, na.rm = TRUE)))
+        fd <- fd[do.call(order, fd), ]
+        rownames(fd) <- NULL
+    }
+    fd
+}
+
+#' Convert the mzML polarity information into a character (+, -, NA)
+#' representing the polarity
+#'
+#' @noRd
+#'
+#' @md
+#'
+#' @author Johannes Rainer
+.polarity_char <- function(x) {
```

Comment:
Is there a unit test for this function? Is the output `c(NA, "-", "+")` expected for `c(-1, 0, 1)`? (see comment above)

Reply:
No, no unit test (yet - I'll add) and yes, the function is expected to return `"+"` for `1`, `"-"` for `0` and `NA` for `-1`. This function is only used to create IDs for SRM/MRM chromatograms similar to the one that is reported in the `mzML` file.

Reply:
I've added now the unit test for the function (https://github.com/lgatto/MSnbase/pull/293/commits/824766ee268a70fb5c3b2a848ae76be8207916e0)